relay-server/src/utils/tracked_future.rs

```diff
@@ -0,0 +1,71 @@
+use futures::{sync::mpsc::Sender, Async, Future, Poll};
+
+/// Message send on the notification channel when the tracked future finishes or is disposed.
+pub struct TrackedFutureFinished;
+
+pub trait IntoTracked<F>
+where
+    F: Future,
+{
+    fn to_tracked(self, notifier: Sender<TrackedFutureFinished>) -> TrackedFuture<F>;
+}
+
+pub struct TrackedFuture<F> {
+    notified: bool,
+    notifier: Sender<TrackedFutureFinished>,
+    inner: F,
+}
+
+impl<F> IntoTracked<F> for F
+where
+    F: Future,
+{
+    fn to_tracked(self, notifier: Sender<TrackedFutureFinished>) -> TrackedFuture<F> {
+        TrackedFuture {
+            notified: false,
+            inner: self,
+            notifier,
+        }
+    }
+}
+
+impl<F> Future for TrackedFuture<F>
+where
+    F: Future,
+{
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let ret_val = self.inner.poll();
+
+        match ret_val {
+            Ok(Async::NotReady) => {}
+            _ => {
+                // future is finished notify channel
+                self.notified = true;
+                self.notifier
+                    .try_send(TrackedFutureFinished)
+                    .map_err(|_| {
+                        log::error!("TrackedFuture could not notify completion");
+                    })
+                    .ok();
```

Comment:
_nit_: Deduplicate in a function:

```rust
fn notify(&mut self) {
    if self.notified { return; }

    self.notified = true;
    self.notifier
        .try_send(TrackedFutureFinished)
        .map_err(|_| {
            log::error!("TrackedFuture could not notify completion");
        })
        .ok();
}
```

Also: Please add a comment explaining why you eagerly run this here and not only in drop.

Reply:
good point.